File: src/shesasyncer/engines/native.py
"""ShesASyncer's native, model-agnostic singing alignment engine.

Unlike a SOFA wrapper, this engine owns the decoding contract. A host supplies
an acoustic scorer; ShesASyncer performs G2P, monotonic phoneme decoding, and
reconstruction of line evidence.
"""

from collections.abc import Callable, Sequence

from ..evidence.acoustic import AcousticFrame, viterbi_phoneme_alignment
from ..lyrics.g2p import G2PEngine
from .adapters import TimedSegment


class NativeSingingEngine:
    name = "shesasyncer-native"

    def __init__(
        self,
        g2p: G2PEngine,
        acoustic_runner: Callable[[str, Sequence[str], str | None], Sequence[AcousticFrame]] | None = None,
    ):
        self.g2p = g2p
        self.acoustic_runner = acoustic_runner

    def available(self) -> bool:
        return self.g2p.available() and callable(self.acoustic_runner)

    def align(
        self,
        audio_path: str,
        lyrics: Sequence[str] | None = None,
        language: str | None = None,
    ) -> tuple[TimedSegment, ...]:
        if not self.available():
            return ()

        lines = list(lyrics or ())
        line_phonemes = [self.g2p.convert(line, language) for line in lines]
        flat = [phoneme for line in line_phonemes for phoneme in line]
        if not flat:
            return ()

        frames = tuple(self.acoustic_runner(audio_path, flat, language))
        spans = viterbi_phoneme_alignment(flat, frames)
        if not spans:
            return ()

        result: list[TimedSegment] = []
        offset = 0
        for line, phonemes in zip(lines, line_phonemes):
            indices = set(range(offset, offset + len(phonemes)))
            line_spans = [span for span in spans if span["token_index"] in indices]
            offset += len(phonemes)
            if not line_spans:
                continue
            result.append(TimedSegment(
                start=float(line_spans[0]["start"]),
                end=float(line_spans[-1]["end"]),
                text=line,
                confidence=sum(float(x["confidence"]) for x in line_spans) / len(line_spans),
                phonemes=tuple({k: v for k, v in span.items() if k != "token_index"} for span in line_spans),
            ))
        return tuple(result)

File: src/shesasyncer/engines/native.js
import { viterbiPhonemeAlignment } from '../evidence/acoustic.js';
import { G2PEngine } from '../lyrics/g2p.js';
import { TimedSegment } from './adapters.js';

export class NativeSingingEngine {
  name = 'shesasyncer-native';
  constructor(g2p, acousticRunner = null) { this.g2p = g2p; this.acousticRunner = acousticRunner; }
  available() { return this.g2p instanceof G2PEngine ? this.g2p.available() && typeof this.acousticRunner === 'function' : typeof this.acousticRunner === 'function' && this.g2p?.available?.(); }
  async align(audioPath, lyrics = [], language = null) {
    if (!this.available()) return [];
    const lines = [...lyrics], linePhonemes = lines.map(line => this.g2p.convert(line, language));
    const flat = linePhonemes.flat(); if (!flat.length) return [];
    const frames = await this.acousticRunner(audioPath, flat, language);
    const spans = viterbiPhonemeAlignment(flat, frames); if (!spans.length) return [];
    const result = []; let offset = 0;
    for (let i = 0; i < lines.length; i++) {
      const phonemes = linePhonemes[i], indices = new Set(Array.from({ length: phonemes.length }, (_, j) => offset + j));
      const lineSpans = spans.filter(span => indices.has(span.tokenIndex)); offset += phonemes.length;
      if (!lineSpans.length) continue;
      result.push(new TimedSegment(lineSpans[0].start, lineSpans.at(-1).end, lines[i], lineSpans.reduce((s, x) => s + x.confidence, 0) / lineSpans.length, [], lineSpans.map(({ tokenIndex, ...x }) => x)));
    }
    return result;
  }
}
